src/game/client/ui.cpp

```diff
@@ -149,284 +210,640 @@ void CUI::UpdateClipping()
+	Value = pScale->ToAbsolute(DoScrollbarH(pID, &ScrollBar, pScale->ToRelative(Value, 0, Max)), 0, Max);
+
+	if(HotItem() != pID && MouseHovered(pRect) && MouseButtonClicked(0))
+		Value = (Value + 1) % NumLabels;
+
+	*pOption = clamp(Value, 0, Max);
+}
+
+float CUI::DrawClientID(float FontSize, vec2 CursorPosition, int ID,
+							const vec4& BgColor, const vec4& TextColor)
+{
+	if(!m_pConfig->m_ClShowUserId)
+		return 0;
+
+	char aBuf[4];
+	str_format(aBuf, sizeof(aBuf), "%d", ID);
+
 	static CTextCursor s_Cursor;
 	s_Cursor.Reset();
 	s_Cursor.m_FontSize = FontSize;
-	s_Cursor.m_MaxWidth = pRect->w;
-	s_Cursor.MoveTo(pRect->x, pRect->y);
+	s_Cursor.m_Align = TEXTALIGN_CENTER;
 
+	vec4 OldColor = TextRender()->GetColor();
 	TextRender()->TextColor(TextColor);
-	const char *pMatch = pHighlighted && pHighlighted[0] ? str_find_nocase(pText, pHighlighted) : 0;
-	if(pMatch)
+	TextRender()->TextDeferred(&s_Cursor, aBuf, -1);
+	TextRender()->TextColor(OldColor);
+
+	const float LinebaseY = CursorPosition.y + s_Cursor.BaseLineY();
+	const float Width = 1.4f * FontSize;
+
+	CUIRect Rect;
+	Rect.x = CursorPosition.x;
+	Rect.y = LinebaseY - FontSize + 0.15f * FontSize;
+	Rect.w = Width;
+	Rect.h = FontSize;
+	Rect.Draw(BgColor, 0.25f * FontSize);
+
+	s_Cursor.MoveTo(Rect.x + Rect.w / 2.0f, CursorPosition.y);
+	TextRender()->DrawTextPlain(&s_Cursor);
+
+	return Width + 0.2f * FontSize;
+}
+
+void CUI::DoTooltip(const void *pID, const CUIRect *pRect, const char *pText)
+{
+	if(MouseHovered(pRect))
 	{
-		TextRender()->TextDeferred(&s_Cursor, pText, (int)(pMatch - pText));
-		TextRender()->TextColor(HighlightColor);
-		TextRender()->TextDeferred(&s_Cursor, pMatch, str_length(pHighlighted));
-		TextRender()->TextColor(TextColor);
-		TextRender()->TextDeferred(&s_Cursor, pMatch + str_length(pHighlighted), -1);
+		m_pActiveTooltip = pID;
+		m_TooltipAnchor = *pRect;
+		str_copy(m_aTooltipText, pText, sizeof(m_aTooltipText));
 	}
-	else
-		TextRender()->TextDeferred(&s_Cursor, pText, -1);
+	else if(m_pActiveTooltip == pID)
+	{
+		m_pActiveTooltip = 0;
+		m_aTooltipText[0] = '\0';
+	}
+}
 
+void CUI::RenderTooltip()
+{
+	if(!m_pActiveTooltip || !m_aTooltipText[0])
+		return;
+	const int64 Now = time_get();
+	static const void *s_pLastTooltip = 0;
+	static int64 s_TooltipActivationStart = 0;
+	if(s_pLastTooltip != m_pActiveTooltip)
+	{
+		// Reset the fade in timer if a new tooltip got active
+		s_TooltipActivationStart = Now;
+		s_pLastTooltip = m_pActiveTooltip;
+	}
+
+	const float SecondsBeforeFadeIn = 0.75f;
+	const float SecondsSinceActivation = (Now - s_TooltipActivationStart)/(float)time_freq();
+	if(SecondsSinceActivation < SecondsBeforeFadeIn)
+		return;
+	const float SecondsFadeIn = 0.25f;
+	const float AlphaFactor = SecondsSinceActivation < SecondsBeforeFadeIn+SecondsFadeIn ? (SecondsSinceActivation-SecondsBeforeFadeIn)/SecondsFadeIn : 1.0f;
+	const CUIRect *pScreen = Screen();
+
+	static CTextCursor s_Cursor;
+	s_Cursor.Reset();
```

Comment:
The text might also change thou, so this would have to do a strcmp. For example the seekbar tooltip changes depending on position.

Reply:
I see. maybe have an extra parameter for `RenderTooltip` that can be passed as StringVersion. if it is not passed, we fallback to `m_pActiveTooltip`.
Or maybe a bool to disable optimization altogether for tooltips that are expected to change?

Generally, I don't do strcmp because I'm afraid of longer static string being compared each frame which is probably O(strlen).

Reply:
Would be possible but I'm not sure it's worth it. Do you also want to change all the DoLabel calls at some point? There are a lot more of those and none reuse the cursor.

Reply:
Definitely, but I have almost no clue on how to do it without flooding the menu code with more static cursors.

Menu text usually only contains a single localized string pointer. Maybe there is a way to setup a cursor per localized string pointer. 

For those displaying numbers or regularly changing we can have a textcursor like we do it normally which isn't too many cases I would imagine.

Browser and ListBox should be optimized individually. And I would probably halt browser optimization until we decided what to do with it's behavior in #2833.

Reply:
I guess tooltip is sparse enough to leave it as-is.

DoLabel can impact performance in-game with in-game menu. But players doesn't need super good performance with the menu open. I would love to do it at some point but it probably wouldn't be my top priority.

Reply:
Yeah, I would also like to avoid clutting the code with text cursors for each UI label.

I presume caching all these cursors would also increase memory usage significantly.